### ISSUE TITLE
System.OutOfMemoryException on Intel GPUs

### ISSUE THREAD

**Issue by bancog52** [HearthSim/Hearthstone-Deck-Tracker#4596]
Hearthstone HDT plugin frequently crashes, the error window cannot be opened. Checking the Crash Reports file, the error message shows:System.OutOfMemoryException: There is not enough memory to continue the execution of the program.
Error status: During plugin operation, it suddenly freezes and requires a restart. After running for two or three rounds, it freezes again. Reinstalling HDT, Hearthstone, setting permissions, running antivirus, updating .NET, and even reinstalling the system have all been ineffective. 
<img width="411" alt="图片1" src="https://github.com/user-attachments/assets/83ccfeb3-b146-463c-af5b-4779b86ec6f1">
I checked the Windows system logs
![图片3](https://github.com/user-attachments/assets/6284cf89-fd8b-4353-882a-43b537d4cfcc)
![图片4](https://github.com/user-attachments/assets/af1e7e43-131a-4858-ad02-dd6dcd498431)

Then I analyzed the possible causes
1、Insufficient system memory or improper application memory management. Ensure the system has enough available memory, close unnecessary applications
2、The exception occurs during WPF rendering process, possibly related to graphics resource management
3、Check for memory leaks or unnecessary resource allocation
4、Make sure you're using the latest version of the software

Finally, after various attempts, I discovered it was a WPF   issue.
By disabling hardware acceleration, the software no longer uses the GPU but instead uses the CPU for rendering, and I found this resolved the issue.
![image](https://github.com/user-attachments/assets/04b75f78-148c-4d4a-9421-77477a094e9e)

I found that many people have encountered this issue but haven't solved it. I'm not sure if this solution will work for everyone. I wonder if this problem can ultimately be resolved at the code level.

**Comment by beheh** [comment]
Hi there, thanks for the report and the investigation!

I don't we can generally propose or automatically disable hardware acceleration, but this seems like it's indicative of some kind of behaviour in WPF, either because it's a bug or because HDT is triggering a bad behaviour. Nonetheless I'll see if we can add this to some FAQ or an error dialog.

**Comment by riQQ** [comment]
Are you using an Intel iGPU?
dotnet/wpf#7704

**Comment by bancog52** [comment]
> 您使用的是 Intel iGPU 吗？ [dotnet/WPF#7704](https://github.com/dotnet/wpf/issues/7704)

Yes, I am using an Intel i5-12500T CPU without a dedicated graphics card.
![image](https://github.com/user-attachments/assets/e1ea87b9-4e1d-44fe-a58d-55cb3519f506)

**Comment by beheh** [comment]
Hello,

I have a test build with one fix that might address this issue (even when removing the registry key again). There's more fixes and performance improvements on the way, but I'd like to get an early lead on whether this addresses the issue:

https://drive.google.com/file/d/1QJPOH6WTlq9bWuF5VKmKJGTwtXv0gdZM/view?usp=sharing

How to use:
1. Download the Zip and unpack it anywhere (do not copy it over any existing install).
2. Run HearthstoneDeckTracker.exe. You may need to explicitly tell Windows to allow this version, as it is not an official release yet.
3. Play as usual. You hopefully don't get a crash anymore.

If you use the build, I'd be happy to hear how it goes (crash or no crash).

**Comment by bancog52** [comment]
> Hello,
> 
> I have a test build with one fix that might address this issue (even when removing the registry key again). There's more fixes and performance improvements on the way, but I'd like to get an early lead on whether this addresses the issue:
> 
> https://drive.google.com/file/d/1QJPOH6WTlq9bWuF5VKmKJGTwtXv0gdZM/view?usp=sharing
> 
> How to use:
> 
> 1. Download the Zip and unpack it anywhere (do not copy it over any existing install).
> 2. Run HearthstoneDeckTracker.exe. You may need to explicitly tell Windows to allow this version, as it is not an official release yet.
> 3. Play as usual. You hopefully don't get a crash anymore.
> 
> If you use the build, I'd be happy to hear how it goes (crash or no crash).

[report.zip](https://github.com/user-attachments/files/18585860/report.zip)

Thank you for providing the test build. I followed the instructions and tested this version, but unfortunately the issue still persists

**Comment by beheh** [comment]
I have a suspicion at this point that this instant crash only affects Intel integrated GPUs, and you can work around there with the registry fix suggested above. If you're a user who encounters more or less instant crashes on HDT v1.38.4+ and you're not on an Intel GPU, please also report here.

**Comment by thesh0ck** [comment]
I have been having this crashing issue with the release client as well (Laptop with integrated Intel graphics). I tried the test build and it does not crash at all.

**Comment by beheh** [comment]
The changes from the test build have since made it into "normal" HDT, weeks ago. So if you're still experiencing them there, then something else may be going on (or maybe the debug build doesn't trigger it).

**Comment by thesh0ck** [comment]
> The changes from the test build have since made it into "normal" HDT, weeks ago. So if you're still experiencing them there, then something else may be going on (or maybe the debug build doesn't trigger it).

Tried the test patch 1 week after it was posted. It worked fine. Went back to the normal build, it did NOT work fine. The newest patch after my post did fix the problem.

**Comment by hsburner** [comment]
Hello,

I'm not sure if I should post here or make a new thread, but I'm having this exact issue, in both battlegrounds and standard. The same crash reports (System.outOfMemoryException), and I also tried the older version linked above. I've tried uninstalling several times.

**Comment by mux** [comment]
For what it's worth, I have tried forcing HDT to use the discrete GPU (RTX 4060) and it did not fix the issue.

**Comment by SigThorarins** [comment]
Just a quick update, I've recently updated my computer from 32GB RAM to 64GB RAM, and it has not entirely solved the problem but it happens a lot less frequently now.
